lib/ansible/modules/network/nuage/nuage_vspk.py

```diff
@@ -0,0 +1,1053 @@
+        :param entity_class: The class of the entity
+        :param parent: The parent on which to add the entity as a member
+        :param set_output: If set to True, sets the Ansible result variables
+        """
+        members = []
+        for member in entity_fetcher.get():
+            if member.id != entity.id:
+                members.append(member)
+        try:
+            parent.assign(members, entity_class)
+        except BambouHTTPError as error:
+            self.module.fail_json(msg='Unable to remove entity as a member: {0}'.format(error))
+        self.result['changed'] = True
+        if set_output:
+            self.result['id'] = entity.id
+            self.result['entities'].append(entity.to_dict())
+
+    def _create_entity(self, entity_class, parent, properties):
+        """
+        Creates a new entity in the parent, with all properties configured as in the file
+        :param entity_class: The class of the entity
+        :param parent: The parent of the entity
+        :param properties: The set of properties of the entity
+        :return: The entity
+        """
+        entity = entity_class(**properties)
+        try:
+            parent.create_child(entity)
+        except BambouHTTPError as error:
+            self.module.fail_json(msg='Unable to create entity: {0}'.format(error))
+        self.result['changed'] = True
+        return entity
+
+    def _save_entity(self, entity):
+        """
+        Updates an existing entity
+        :param entity: The entity to save
+        :return: The updated entity
+        """
+        try:
+            entity.save()
+        except BambouHTTPError as error:
+            self.module.fail_json(msg='Unable to update entity: {0}'.format(error))
+        self.result['changed'] = True
+        return entity
+
+    def _delete_entity(self, entity):
+        """
+        Deletes an entity
+        :param entity: The entity to delete
+        """
+        try:
+            entity.delete()
+        except BambouHTTPError as error:
+            self.module.fail_json(msg='Unable to delete entity: {0}'.format(error))
+        self.result['changed'] = True
+
+    def _wait_for_job(self, entity):
+        """
+        Waits for a job to finish
+        :param entity: The job to wait for
+        """
+        running = False
+        if entity.status == 'RUNNING':
+            self.result['changed'] = True
+            running = True
+
+        while running:
+            time.sleep(1)
+            entity.fetch()
+
+            if entity.status != 'RUNNING':
+                running = False
+
+        self.result['entities'].append(entity.to_dict())
+        if entity.status == 'ERROR':
+            self.module.fail_json(msg='Job ended in an error')
+
+
+def main():
+    """
+    Main method
+    """
+    module = AnsibleModule(
+        argument_spec=dict(
+            auth=dict(
+                required=True,
+                type='dict',
+                no_log=True,
```

Comment:
@gundalow while doing negative tests, i ran into an issue in the current `module_utils/basic.py` code, namely:
https://github.com/ansible/ansible/blob/757758cd1a09beb7b99aef7b5dab04bb2f4fc428/lib/ansible/module_utils/basic.py#L1739
```
value = self.params[k]
```

This will fail on when doing the check of the sub options, with something like:
```
Traceback (most recent call last):
  File "/var/folders/7t/j48vp3pd2fzbz30yr8xjxpr00000gq/T/ansible_yoy2Ti/ansible_module_nuage_vspk.py", line 1052, in <module>
    main()
  File "/var/folders/7t/j48vp3pd2fzbz30yr8xjxpr00000gq/T/ansible_yoy2Ti/ansible_module_nuage_vspk.py", line 1041, in main
    supports_check_mode=True
  File "/var/folders/7t/j48vp3pd2fzbz30yr8xjxpr00000gq/T/ansible_yoy2Ti/ansible_modlib.zip/ansible/module_utils/basic.py", line 759, in __init__
  File "/var/folders/7t/j48vp3pd2fzbz30yr8xjxpr00000gq/T/ansible_yoy2Ti/ansible_modlib.zip/ansible/module_utils/basic.py", line 1759, in _check_argument_types
  File "/var/folders/7t/j48vp3pd2fzbz30yr8xjxpr00000gq/T/ansible_yoy2Ti/ansible_modlib.zip/ansible/module_utils/basic.py", line 1739, in _check_argument_types
KeyError: 'api_enterprise'
```

The code should be:
```
value = param[k]
```

(In case you are interested, my negative test cases can be found here: https://github.com/nuagenetworks/vspk-ansible/blob/beccbf3269a3dd5163422814506e9a4b5b244240/roles/nuage-vspk-negative-tests/tasks/main.yml )